lib/parallel_test_base.pm

```diff
@@ -0,0 +1,1946 @@
+    my $_localarch = script_output("uname -i");
+    my $_peerip = get_var('PEER_IPADDR');
+    my $_peerarch = script_output("ssh -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no root\@$_peerip uname -i");
+    save_screenshot;
+    croak("Architecture $_localarch on $_localip does not match $_peerarch on $_peerip") if ($_localarch ne $_peerarch);
+}
+
+=head2 check_host_os
+
+Check source and destination hosts operating system version. Guest migration can
+not be done from the newer to the older. Main argument is host role (src or dst).
+=cut
+
+sub check_host_os {
+    my ($self, %args) = @_;
+    $args{role} //= 'src';
+
+    record_info("Check host os");
+    my $_ret = 0;
+    my $_localip = get_var('LOCAL_IPADDR');
+    my $_peerip = get_var('PEER_IPADDR');
+    if (is_sle) {
+        my ($_localosver, $_localossp,) = get_os_release;
+        my ($_peerosver, $_peerossp,) = get_os_release("ssh -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no root\@$_peerip");
+        print "LOCAL OS $_localosver-sp$_localossp PEER OS $_peerosver-sp$_peerossp\n";
+        save_screenshot;
+        unless (($_peerosver > $_localosver) or ($_peerosver == $_localosver and $_peerossp >= $_localossp)) {
+            $_ret = 1;
+            if ($args{role} eq 'src') {
+                croak("Destination os $_peerosver-sp$_peerossp falls behind source os $_localosver-sp$_localossp");
+            }
+            elsif ($args{role} eq 'dst') {
+                record_info("Source os $_peerosver-sp$_peerossp  falls behind destination os $_localosver-sp$_localossp");
+            }
+        }
+    }
+    return $_ret;
+}
+
+=head2 check_host_virtualization
+
+Check virtualization modules and services are ready, otherwise test run can not
+proceed.
+=cut
+
+sub check_host_virtualization {
+    my $self = shift;
+
+    record_info("Check host virtualization");
+    if (is_kvm_host) {
+        assert_script_run("lsmod | grep kvm");
+    }
+    elsif (is_xen_host) {
+        assert_script_run("lsmod | grep xen");
+    }
+
+    if (script_run("systemctl is-active libvirtd") != 0) {
+        systemctl("stop libvirtd", ignore_failure => 1);
+        systemctl("start libvirtd");
+        systemctl("is-active libvirtd");
+    }
+    save_screenshot;
+}
+
+=head2 check_host_network
+
+Check source and destination hosts are in the same subnet, otherwise test run
+can not proceed.
+=cut
+
+sub check_host_network {
+    my $self = shift;
+
+    record_info("Check host network");
+    my $_localgw = script_output("ip route show default | awk \'{print \$3}\'");
+    my $_localip = get_var('LOCAL_IPADDR');
+    my $_peerip = get_var('PEER_IPADDR');
+    my $_peergw = script_output("ssh -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no root\@$_peerip ip route show default | awk \'{print \$3}\'");
+    save_screenshot;
+    croak("$_localip and $_peerip hosts are not in the same subnet") unless ($_localgw eq $_peergw);
```

Comment:
Would you please share where this requirement comes? I do not remember any on this for migration. In real migration, I think it is possible that src and dst are on different sub-network.

Reply:
https://susedoc.github.io/doc-sle/main/single-html/SLES-virtualization/#sec-libvirt-admin-migrate
```Host and target machine should be in the same subnet on the network, otherwise networking will not work after the migration. ```

Reply:
Yes, that's true. My understanding is that the src and dst should be both in the network that the migrated vm is created in/used in its network config,  but it is not necessarily the default network for src and dst?  Otherwise, it won't say "otherwise networking will not work AFTER the migration" ?

Reply:
Do you mean "the same subnet" refers to the network in which src/dst guest reside ?

Reply:
That's my understanding. As long as src and dst host can reach each other, the vm will be able to be migrated from src to dst. Then if dst has the same network that the vm xml requires in its network configuration, which also applied on src host, the vm's networking will work after migration.

Reply:
I can remove this checking if you think this is not very necessary. But I think in the same subnet is reasonable.

Reply:
I know it takes effort to implement and test the code for this. Sorry for suggesting to remove. Among our pair machines on OSD, there is one pair won't match same subnet, scooter is moved to new FC lab while the other one is in old NUE lab. It can also save some test time.